### PR TITLE
[#1 Methodological issues fix]

### LINKED ISSUE #1: Methodological issues in ML pipeline
### Overview

Went over the notebook and noticed several issues in methodology:

🚨 CRITICAL ISSUES IDENTIFIED
1. MAJOR LOOK-AHEAD BIAS: Feature-Label Alignment
Severity: CRITICAL ❌
Location: Data preparation section (cell with X = sample_df.drop(['label'], axis=1))
Problem:
pythonX = sample_df.drop(['label'], axis=1)
y = sample_df['label']

# Simple chronological split
X_train = X.iloc[:split_idx]
X_test = X.iloc[split_idx:]
y_train = y.iloc[:split_idx]
y_test = y.iloc[split_idx:]
Issue: You're using features at time t to predict labels at time t, but the label at time t is created using future prices from t+1 to t+20. This is classic look-ahead bias.
Impact:

Your models are learning from future information that wouldn't be available at prediction time
Real trading performance will be significantly worse than backtested results
All accuracy metrics are artificially inflated

Fix:
python# Shift labels forward by 1 bar (predict NEXT bar's outcome)
sample_df['target'] = sample_df['label'].shift(-1)
sample_df = sample_df.dropna(subset=['target'])

X = sample_df.drop(['label', 'target'], axis=1)
y = sample_df['target']

# Now when you split:
# Features at time t predict label at time t+1
# Label at t+1 uses prices from t+2 to t+21
# This is valid because at time t, we don't know t+1 yet

2. CRITICAL: Feature Leakage - Price-Based Features
Severity: CRITICAL ❌
Location: Feature engineering and feature selection
Problem:
pythonfeatures_to_scale = [
    'close',      # ❌ Using current close price
    'ema_10',     # ❌ Includes current bar
    'sma_50',     # ❌ Includes current bar
    # ...
]
Issue: You're including close, ema_10, and sma_50 which all incorporate the current bar's closing price. When predicting whether to enter a trade, you wouldn't have the close price yet.
Impact:

Model has access to information that wouldn't be available at decision time
You'd need to enter trades AFTER the bar closes, introducing execution delay
True performance will be worse due to slippage and timing issues

Fix:
python# Option 1: Remove current-price features entirely
features_to_drop = ['close', 'open', 'high', 'low']

# Option 2: Use LAGGED values only
df['close_lag1'] = df['close'].shift(1)  # Previous bar's close
df['ema_10_lag1'] = df['ema_10'].shift(1)

# Option 3: Use features that are calculated BEFORE bar close
# (e.g., mid-bar indicators if you're trading intrabar)

3. CRITICAL: No Embargo Period in Cross-Validation
Severity: HIGH ⚠️
Location: TimeSeriesSplit implementation
Problem:
pythontscv = TimeSeriesSplit(n_splits=5)

for fold, (train_idx, test_idx) in enumerate(tscv.split(X_train_scaled)):
    # No purging or embargo
    X_fold_train = X_train_scaled.iloc[train_idx]
    X_fold_val = X_train_scaled.iloc[val_idx]
Issue: Since triple barrier labels can span up to 20 bars:

Bar 100's label uses prices up to bar 120
Bar 119's label uses prices up to bar 139
If train set ends at bar 119 and test starts at bar 120, there's overlap

Impact:

Training data leaks into test data through overlapping labels
CV scores are optimistically biased
Model appears better than it actually is

Fix:
pythondef get_purged_train_test_idx(train_idx, test_idx, embargo_bars=20):
    """Remove training samples that overlap with test period."""
    test_start = test_idx[0]
    # Remove samples within embargo_bars of test start
    train_idx = train_idx[train_idx < test_start - embargo_bars]
    return train_idx, test_idx

# In CV loop:
for fold, (train_idx, val_idx) in enumerate(tscv.split(X_train_scaled)):
    train_idx, val_idx = get_purged_train_test_idx(train_idx, val_idx, embargo_bars=20)
    # ... rest of code

4. Data Leakage: Scaler Fitted on Full Training Set
Severity: MEDIUM ⚠️
Location: Pre-CV scaling
Problem:
python# Fit scaler on full training data BEFORE CV
scaler = RobustScaler()
X_train_scaled = X_train.copy()
X_train_scaled[features_to_scale] = scaler.fit_transform(X_train[features_to_scale])

# Then use in CV loop...
for fold, (train_idx, val_idx) in enumerate(tscv.split(X_train_scaled)):
    # Uses pre-scaled data
Issue: The scaler has "seen" the validation fold data during the initial fit_transform, causing minor leakage.
Impact:

Slight optimistic bias in CV results
Not as severe as other issues but violates best practices

Fix: ✅ You actually DO fix this correctly later in your CV loop!
python# Inside CV loop - CORRECT approach (which you do have):
fold_scaler = RobustScaler()
X_fold_train_scaled[features_to_scale] = fold_scaler.fit_transform(
    X_fold_train[features_to_scale]
)
X_fold_val_scaled[features_to_scale] = fold_scaler.transform(
    X_fold_val[features_to_scale]
)
Verdict: This is handled correctly in your CV implementation ✅

5. Consecutive Bars Features - Potential Forward-Looking
Severity: MEDIUM ⚠️
Location: Regime features
Problem:
pythondf['consecutive_up'] = df['positive_returns'].groupby(
    (df['positive_returns'] != df['positive_returns'].shift()).cumsum()
).cumsum()
Issue: This counts consecutive up/down bars including the current bar. At the time of prediction, you don't know if the current bar will be up or down yet.
Impact:

Minor look-ahead if used without proper offsetting
Feature incorporates current bar's return which isn't known until close

Fix:
python# Use lagged consecutive counts
df['consecutive_up'] = df['positive_returns'].groupby(
    (df['positive_returns'] != df['positive_returns'].shift()).cumsum()
).cumsum().shift(1)  # Look at consecutive count UP TO previous bar

6. Missing: Forward-Looking Feature Check
Severity: MEDIUM ⚠️
Location: Feature engineering section
Problem: No explicit verification that all features use only past data.
Features to audit:

returns = df['close'].pct_change() ✅ Good (uses t and t-1)
cvd = df['volume_delta'].cumsum() ✅ Good (cumulative through current)
rsi_14 ✅ Good (rolling calculation on past data)
macd ✅ Good (EMA uses past data)
All rolling features ✅ Good (look backward only)

Verdict: Your technical features are actually fine! ✅

7. Class Imbalance Not Fully Addressed
Severity: LOW ℹ️
Location: Model training
Observation: You use class_weight='balanced' for many models, which is good. However:
python# Label distribution check needed
print(df['label'].value_counts())

# Expected distribution:
# 1 (profit):  ~15-30% (depending on profit_pct)
# 0 (time):    ~40-60% (most common)
# -1 (loss):   ~15-30% (depending on stop_pct)
Recommendations:

Consider SMOTE or other resampling for extreme imbalance
Use stratified sampling if possible (though TimeSeriesSplit doesn't support it)
Focus on precision/recall for minority classes (profit/loss) over overall accuracy

8. Triple Barrier Implementation - Edge Cases
Severity: LOW ℹ️
Location: add_triple_barrier_labels function
Potential issues:
python# Issue 1: What if high AND low breach barriers in same bar?
if df['high'].iloc[j] >= upper_barrier:
    barrier_hit = 'upper'
    break

if df['low'].iloc[j] <= lower_barrier:
    barrier_hit = 'lower'
    break
Problem: If both barriers are hit in the same bar, you assume upper is hit first. In reality, you'd need tick data to know the order.
Impact: Minor - probably doesn't matter much in practice, but could affect ~1-5% of labels.
Consider:
python# More conservative: if both hit, mark as uncertain or use mid-price exit
if df['high'].iloc[j] >= upper_barrier and df['low'].iloc[j] <= lower_barrier:
    barrier_hit = 'both'  # Handle separately
    exit_price = entry_price  # Or use mid-price
    break

📊 CORRECT IMPLEMENTATIONS (WELL DONE!)
✅ TimeSeriesSplit Usage
You correctly use TimeSeriesSplit instead of regular cross-validation, respecting temporal ordering.
✅ Scaler Refitting in CV Folds
You properly refit the scaler in each CV fold:
pythonfold_scaler = RobustScaler()
X_fold_train_scaled[features_to_scale] = fold_scaler.fit_transform(...)
✅ No Shuffling in Train/Test Split
pythonX_train = X.iloc[:split_idx]  # Chronological, no shuffle
✅ Proper Feature Engineering
All your rolling windows, EMAs, and technical indicators correctly look backward only.
✅ Class Weights
You use class_weight='balanced' to handle imbalanced classes.

🔧 PRIORITY FIXES RANKED
1. MUST FIX IMMEDIATELY 🔴

Shift labels forward by 1 bar to align features with labels properly
Remove current bar price features (close, ema_10, sma_50) or lag them by 1 bar

2. FIX BEFORE PRODUCTION 🟡

Add embargo period to TimeSeriesSplit (20 bars based on your time_bars parameter)
Lag consecutive bar features to not include current bar

3. NICE TO HAVE 🟢

Add explicit forward-looking feature checker function
Handle both-barriers-hit edge case in triple barrier
Consider more sophisticated resampling for class imbalance

### PR DESCRIPTION
**Pull request by devinvegas**
Draft

### CODE REVIEW

**Comment by devinvegas** [comment]
Triple barrier, basic implementation and ML pipeline steps added
